addons/binding/org.openhab.binding.amazonechocontrol/src/main/java/org/openhab/binding/amazonechocontrol/internal/Connection.java

```diff
@@ -0,0 +1,925 @@
+
+            if (postLoginData(queryParameters, postData) != null) {
+                throw new ConnectionException(
+                        "Login fails. Check your credentials and try to login with your webbrowser to http(s)://<youropenhab:yourport>/amazonechocontrol/"
+                                + accountThingId);
+            }
+
+        } catch (Exception e) {
+            // clear session data
+            cookieManager.getCookieStore().removeAll();
+            sessionId = null;
+            loginTime = null;
+            verifyTime = null;
+            logger.info("Login failed: {}", e.getLocalizedMessage());
+            // rethrow
+            throw e;
+        }
+    }
+
+    public @Nullable String postLoginData(@Nullable String optionalQueryParameters, @Nullable String postData)
+            throws IOException, URISyntaxException {
+        // build query parameters
+        @Nullable
+        String queryParameters = optionalQueryParameters;
+        if (queryParameters == null) {
+            queryParameters = "session-id=" + URLEncoder.encode(sessionId, "UTF-8");
+        }
+
+        // build referer link
+        String referer = "https://www." + amazonSite + "/ap/signin?" + queryParameters;
+        Map<String, String> headers = new HashMap<String, String>();
+        headers.put("Referer", referer);
+
+        // make the request
+        URLConnection request = makeRequest("POST", "https://www." + amazonSite + "/ap/signin", postData, false, true,
+                headers);
+
+        String response = convertStream(request.getInputStream());
+        logger.debug("Received content after login {}", response);
+
+        String host = request.getURL().getHost();
+        if (!host.equalsIgnoreCase(new URI(alexaServer).getHost())) {
+            return response;
+        }
+        if (response.contains("<title>Amazon Alexa</title>")) {
+            logger.debug("Response seems to be alexa app");
+        } else {
+            logger.info("Response maybe not valid");
+        }
+
+        // verify login
+        if (!verifyLogin()) {
+            return response;
+        }
+        logger.debug("Login succeeded");
+        return null;
+    }
+
+    public boolean verifyLogin() throws IOException, URISyntaxException {
+        String response = makeRequestAndReturnString(alexaServer + "/api/bootstrap?version=0");
+        Boolean result = response.contains("\"authenticated\":true");
+        if (result) {
+            verifyTime = new Date();
+            if (loginTime == null) {
+                loginTime = verifyTime;
+            }
+        }
+        return result;
+    }
+
+    public void logout() {
+        cookieManager.getCookieStore().removeAll();
+        sessionId = null;
+        loginTime = null;
+        verifyTime = null;
+    }
+
+    // parser
+    private <T> T parseJson(String json, Class<T> type) throws JsonSyntaxException {
+        try {
+            Gson gson = new Gson();
```

Comment:
This is potentially expensive - you should instantiate it only once as a field in that class.